docs/proposals/20230407-managed-k8s-capi-contract-changes.md

```diff
@@ -0,0 +1,556 @@
+
+#### Story 5
+
+As a Cluster API provider developer, I want to enable the ClusterClass feature for a Managed Kubernetes service, so that users can take advantage of an improved UX with ClusterClass-based clusters.
+
+#### Story 6
+
+As a cluster operator, I want to use Cluster API to provision and manage the lifecycle of worker nodes that utilizes my cloud providers' managed instances (if they support them), so that I don't have to worry about the management of these instances.
+
+#### Story 7
+
+As a service provider I want to be able to offer Managed Kubernetes clusters by using CAPI referencing my own managed control plane implementation that satisfies Cluster API contracts.
+
+### Design
+
+Below we are documenting API changes and controllers changes required to omit the `<Infra>Cluster` entirely and to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint.
+
+#### Core Cluster API changes
+
+This proposal does not introduce any breaking changes for the existing "core" API. More specifically:
+
+The existing Cluster API types are already able to omit the `<Infra>Cluster`:
+
+- The `infrastructureRef` field on the Cluster object is already a pointer and thus it could be set to nil, and in fact we are already creating Clusters without `infrastructureRef` when we use a cluster class).
+- The `infrastructure.Ref` field on the ClusterClass objects already a pointer and thus it could be set to nil, but in this case it is required to change the validation webhook to allow the user to not specify it; on top of that, when validating inline patches, we should reject patches targeting the infrastructure template objects if not specified.
+
+In order to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint we are going to introduce a new `ClusterEndpoint` CRD, below some example:
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterEndpoint
+spec:
+  hostname: 'name-1234567890.region.elb.amazonaws.com'
+  port: 1234
+  type: ExternalControlPlaneEndpoint
+```
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterEndpoint
+spec:
+  ip: 10.40.85.102
+  port: 1234
+  type: ExternalControlPlaneEndpoint
+```
+
+This type will be the primary way for both `<Infra>Cluster` or `ControlPlane Provider` to communicate the ClusterAddress to "core" Cluster API controllers.
+
+The `<Infra>Cluster` object which is currently using the `spec.controlPlaneEndpoint` for the same scope will continue to work because "core" Cluster API controllers will continue to recognize when this field is set and take care of generating the `ClusterEndpoint` automatically; however this mechanism should be considered as a temporary machinery to migrate to the new CRD, and it will be removed in future versions of Cluster API.
```

Comment:
Q: Should we eventually also drop the controlPlaneEndpoint field in the Cluster CR? Seems like redundant information once we have the ClusterEndpoint CRD

Reply:
Added some language to that effect, PTAL.

Reply:
(I'm not entirely sure if we might want to keep the endpoint exposed on the Cluster-level for easier consumption by users/other tools)

Reply:
I think we can defer the decision about dropping the controlPlaneEndpoint field in the Cluster CR to the work that will introduce more control plan endpoints (or to the v1beta2) work